### DISCUSSION TITLE
How to use remote turso db with v0.23.0-rc

### DISCUSSION THREAD

**Discussion by orenaksakal** [pocketbase/pocketbase#5611] [Q&A] [answered]
Congrats for 23-rc release 🚀

I think this will be documented in future, but would love to learn if it will be possible to useturso db remote url with 0.23.0 and how can it be done

https://docs.turso.tech/sdk/go/quickstart#remote-only

**Comment by ganigeorgiev** [comment, marked as answer]
I haven't tested it and I'm not sure what configuration options Turso supports/requires but a minimal example could be:

```go
package main

import (
	"log"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase"

	_ "github.com/tursodatabase/libsql-client-go/libsql"
)

// register the libsql driver to use the same interface implementation as the already existing sqlite3
func init() {
	dbx.BuilderFuncMap["libsql"] = dbx.BuilderFuncMap["sqlite3"]
}

func main() {
	app := pocketbase.NewWithConfig(pocketbase.Config{
		DBConnect: func(dbPath string) (*dbx.DB, error) {
			// register also connection pragmas if needed?
			return dbx.Open("libsql", "...")
		},
		// for all other options (including max allowed open connections), please check
		// https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.23.0-rc2#Config
	})

	// any custom hooks or plugins...
	// (for reference check https://github.com/pocketbase/pocketbase/blob/develop/examples/base/main.go)

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```
(_as a reference of the current used `PRAGMA`s for example you can check the [`db_connect_cgo.go`](https://github.com/pocketbase/pocketbase/blob/0c2266490fea894995cd2e2df3c9ad758d121c00/core/db_connect_cgo.go#L26-L32); note that they may not be supported or needed by Turso_)

Run `go build` and then start the generated PocketBase executable as usual.

**Reply by orenaksakal** [reply]
I'm stuck at 

```bash
2024/10/06 14:27:06 Failed to apply migration 1640988000_init.go: _mfas error: failed to run optimize after the fields changes: failed to execute SQL:
SQL string could not be parsed: unsupported statement: PRAGMA optimize
failed to execute SQL:
Transaction timed out
exit status 1
```

**Reply by ganigeorgiev** [reply]
Either there is a bug in the Turso driver or `PRAGMA optimize` is not supported (most likely the latter).

With v0.23 we run `PRAGMA optimize` once a day and after every schema change per the [SQLite recommendation](https://www.sqlite.org/pragma.html#pragma_optimize):

> ... "PRAGMA optimize" again at periodic intervals - perhaps once per day. All applications should run "PRAGMA optimize" after schema changes, especially [CREATE INDEX](https://www.sqlite.org/lang_createindex.html).

I'm not sure what to recommend here. One possible workaround I guess could be to treat the `PRAGMA optimize` errors as non-critical and only log them with `WARN` severity. I'll consider it before the final v0.23.0 release but there is no guarantee that this may not happen in the future with other `PRAGMA` statements if the driver doesn't support them (e.g. `analysis_limit`, `writable_schema`, `foreign_keys`, etc.).

Eventually you can also try to ask in the Turso support channels what `PRAGMA`s are supported so we can eventually keep them in mind and avoid them during development but I can't promise that we will always comply with their requirements.

**Reply by ganigeorgiev** [reply]
I've pushed a small change in the [`develop` branch](https://github.com/pocketbase/pocketbase/commit/1f7aba40f0a0b1b23d04dc14e5cff0f538b357fb) to make the `PRAGMA optimize` statements optional (_it will only log in case of an error_) but again there is no guarantee that this may not happen with other `PRAGMA` statements.

I haven't tested yet actually Turso and the libsql driver to determine to what extend they differ from the other Go implementations but based on a quick look at some of the opened issues it seems that are indeed some differences that may proven difficult to integrate with PocketBase (or any Go project in general that expects similar behavior to the other existing Go SQLite drivers):
- https://github.com/tursodatabase/go-libsql/issues/28
- https://github.com/tursodatabase/libsql/issues/1553
- https://github.com/tursodatabase/libsql/issues/1119

Technically we could add conditional statements like `if (isLibsqlDriver){ ... } else { ... }` but that doesn't seem maintainable in the long run and it would be better the driver incompatibilities to be resolved in the upstream when possible.

**Reply by orenaksakal** [reply]
that actually worked I've commented out pragma optimize statements locally and built using examples/base 🤯
also I have asked turso about supported pragmas and will report back here when I get an answer.

turso is HA, multi location with PITR

this means I can have multiple pocketbase instances all pointing to same turso db across the world and I won't have to worry about replication anymore, only keeping instances in sync (custom hooks etc) case should be enough am I right? (flyio helps with this a lot), how would realtime api react to this?

![Screenshot 2024-10-07 at 1 33 39 PM](https://github.com/user-attachments/assets/4f546bb7-50b0-4d43-afec-22d4f6e10fdc)

**Reply by orenaksakal** [reply]
noticed log writes are failing with

```bash
2024/10/07 16:20:21 Failed to write log &{{ 87czolwgh8f10mn} 2024-10-07 22:20:18.186Z {"auth":"_superusers","authId":"44d17s51v6z6r7b","details":"failed to execute SQL:\nSQLite error: no such table: _logs; failed query: SELECT {{_logs}}.* FROM \"_logs\" WHERE JSON_EXTRACT([[data]], '$.auth') IS NOT {:tTbi3x} ORDER BY [[_logs]].[[_rowid_]] DESC LIMIT 50","error":"Something went wrong while processing your request.","execTime":305.72075,"method":"GET","referer":"http://127.0.0.1:8090/_/?","remoteIP":"127.0.0.1","status":400,"type":"request","url":"/api/logs?page=1\u0026perPage=50\u0026sort=-%40rowid\u0026skipTotal=1\u0026filter=data.auth!%3D%22_superusers%22","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36","userIP":"127.0.0.1"} GET /api/logs?page=1&perPage=50&sort=-@rowid&skipTotal=1&filter=data.auth!="_superusers" 8} 
failed to execute SQL:
SQLite error: no such table: _logs
```

**Reply by ganigeorgiev** [reply]
PocketBase internally maintains 2 databases - `pb_data/data.db` and `pb_data/logs.db` (with v0.23 the latter is renamed to `pb_data/auxiliary.db`).

The `DBConnection` function is called for each of the databases (see the `dbPath` argument).

The above error *"no such table: _logs"* could be return for many reasons but if you use PocketBase v0.23.0 this is most likely the same as in https://github.com/pocketbase/pocketbase/discussions/5618. Or in other words, it is possible that you've ran the migrations (they are executed automatically on `serve`), deleted the auxiliary database and then started again the application. With the new behavior we keep track of the executed migrations only in `pb_data/data.db`. As mentioned in the linked issue, I'll consider with the final v0.23.0 release to init manually the minimal required aux schema aux in case of the mentioned scenario.

Regarding _"this means I can have multiple pocketbase instances all pointing to same turso db"_ - I don't exactly understand what do you mean by this.

Keep in mind that PocketBase out of the box can scale only vertical. The database is not the limiting factor, the application itself is simply not designed with that use case in mind. Realtime events, or any other hook events, are handled on application level and there is no builtin synchronization mechanism for multiple instances (_we don't rely on db triggers if that are your expectations because they are too limited and we can't control when they fire_).

**Reply by orenaksakal** [reply]
Got it. [the list of turso supported pragmas](https://github.com/avinassh/libsql/blob/4c3f5fc7f/libsql-server/src/query_analysis.rs#L147)

**Reply by tigawanna** [reply]
please share implemention if you get it working 🥺

**Reply by thewandererbg** [reply]
I can get it running like this. Not sure whether it is good.

```
func main() {
	err := godotenv.Load()
	if err != nil {
		log.Println(err)
	}

	app := pocketbase.NewWithConfig(pocketbase.Config{
		DBConnect: func(dbPath string) (*dbx.DB, error) {
			if strings.Contains(dbPath, "data.db") {
				return dbx.Open("libsql", os.Getenv("TS_URL"))
			} else {
				return dbx.Open("sqlite3", dbPath)
			}
		},
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

One error I can find is when I update the collection in the front end (add fields, update API rules ...), I get this error 

```
Failed to update collection. Raw error: 
failed to execute SQL:
SQL string could not be parsed: unsupported statement: PRAGMA writable_schema = ON.
```

**Reply by ganigeorgiev** [reply]
@thewandererbg Regarding the unsupported `writable_schema` PRAGMA - I'll consider before the final release to see if it can be replaced with something else.

Regarding the DBConnect function - I cannot speak about Turso as I haven't really explored their documentation in depth for all the available configuration options but your fallback Open call, which would be applied for the `auxiliary.db`, is not properly configured as the database is not opened in WAL mode and it is missing the `busy_timeout` (see for example the pure Go driver, aka. when build with `CGO_ENABLED=0` - https://github.com/pocketbase/pocketbase/blob/develop/core/db_connect_nocgo.go).

With that said, I see that there is an interest of using Turso but this is not a primary goal of PocketBase and people need to manage their expectations and have clear understanding of why they want to use Turso/libSQL (or any other custom SQLite driver implementation) in the first place before jumping on the hype train. As mentioned in several other discussions, the only benefit from using PocketBase with something like Turso would be the auto replication support (which you could already do with external tools like Litestream) and eventually allowing deploying on hosts with ephemeral storage (aka. when there is no support for persistent disk/volumes). If you expect somehow better performance then most likely you'll be disappointed, especially if you are using the remote db setup because a network request will almost always be slower compared to a direct local file access.

**Reply by thewandererbg** [reply]
@ganigeorgiev thank you for the info. Sure, nothing beats the simplicity and speed of a local embedded SQLite.

**Reply by ganigeorgiev** [reply]
I've pushed a small update in the `develop` branch that removes the `writable_schema` PRAGMA usage.

**Reply by benallfree** [reply]
@ganigeorgiev PocketHost gets asked about replication all the time, but as I see it PocketBase would need something like a `main` and `replicant` mode, and some way to broadcast events as well. It would be a big win to solve this for people, but I'm not sure how to do it without changes to PocketBase itself. Is that how you see it as well?

**Reply by ganigeorgiev** [reply]
@benallfree I don't understand your question and I don't see how it is related to the above discussion. Please open a new one and elaborate what actually is the problem.

**Reply by thewandererbg** [reply]
Another small issue I come across is that if I want to use turso CGO library with "github.com/tursodatabase/go-libsql" it will conflict with github.com/mattn/go-sqlite3 in https://github.com/pocketbase/pocketbase/blob/develop/core/db_connect_cgo.go
even though that was not needed.

When I comment out that package in the file, then I can use go-libsql with no issue.

**Reply by ganigeorgiev** [reply]
@thewandererbg What do you mean by "it will conflict"? Could you paste the exact error?

**Reply by thewandererbg** [reply]
@ganigeorgiev 

Here is the minimal sample code that can lead to that error:

```
package main

import (
	"log"

	"github.com/pocketbase/pocketbase"
	_ "github.com/tursodatabase/go-libsql"
)

func main() {
	app := pocketbase.New()

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

The error was very long. Here is a short part of it.

```
/usr/bin/ld: /home/admin/go/pkg/mod/github.com/tursodatabase/go-libsql@v0.0.0-20241011135853-3effbb6dea5c/lib/linux_amd64/libsql_experimental.a(sqlite3mc.c.o): in function `sqlite3_status64':
sqlite3mc.c:(.text+0x0): multiple definition of `sqlite3_status64'; /tmp/go-link-1407367964/000027.o:/home/admin/go/pkg/mod/github.com/mattn/go-sqlite3@v1.14.24/sqlite3-binding.c:23958: first defined here

....

/usr/bin/ld: /home/admin/go/pkg/mod/github.com/tursodatabase/go-libsql@v0.0.0-20241011135853-3effbb6dea5c/lib/linux_amd64/libsql_experimental.a(sqlite3mc.c.o): in function `sqlite3_db_filename':
sqlite3mc.c:(.text+0x24d50): multiple definition of `sqlite3_db_filename'; /tmp/go-link-1407367964/000027.o:/home/admin/go/pkg/mod/github.com/mattn/go-sqlite3@v1.14.24/sqlite3-binding.c:184655: first defined here
```

**Reply by ganigeorgiev** [reply]
Hm, I'll test it later but from just the error it seems that it is caused because both libraries import the C SQLite amalgamation and if that's the case then I'm not sure what the general solution should be other than put the default CGO driver behind a custom build tag.
Another option could be to pass `--allow-multiple-definition` option to the linker via the ldflags (at least based on the [ld docs](https://linux.die.net/man/1/ld) it should work but I haven't tested it).

**Reply by ganigeorgiev** [reply]
Building with:
```
CGO_ENABLED=1 go build --ldflags='-extldflags "-Wl,--allow-multiple-definition"'
```
seems to build successfully when using the local/embedded-replica libSQL driver but it is more of a hack. I'll consider before the final release changing the current build constraints and loading the default `mattn/go-sqlite3` CGO driver only if a specific build tag is provided (_or just not load it and have an example in the documentation for users to manually register it in case they want to use it_) since it could cause issues with other similar SQLite CGO drivers.

**Reply by ganigeorgiev** [reply]
I've pushed a small update in the `develop` branch that removes the default `mattn/go-sqlite3` registration. Additionally the default db connect function is now also exported as `core.DefaultDBConnect` so that it could eventually be used as a fallback.

**Comment by pedrosimao** [comment]
exciting news!
